### DISCUSSION TITLE
Best place to put library files used throughout one course

### DISCUSSION THREAD

**Discussion by m0n0chr0m3** [dodona-edu/feedback-and-support#36] [Questions - Vragen]
## Summary
I'm preparing a new course repository that will make use of [the Scheme Dodona judge](https://docs.dodona.be/nl/references/judges/#scheme).  Most exercises of the course depend on an `r7rs` Scheme software library. I'm wondering what is the best way to expose libraries to exercises in Dodona.

## Constraints
There are two constraints that influence my options:

- First, the folders containing  libraries need to be be specified at the command-line invocation of the `racket` binary. As far as I know, it is not possible to specify a library's folder dynamically (i.e., from `r7rs` Scheme code). Hence it does not appear to be possible to have an individual "exercise" load a copy of the library from within its`evaluation` folder: the libraries' folders _have_ to be known at the point at which the `racket` binary is invoked. This binary is invoked by the `run` script of the judge, i.e., once per submission.
- Second, the library is course-specific. It would not make sense to simply include every course-specific library in the Scheme Dodona judge, as that would entail that development of each such library would have to take place in the repository containing the judge.

Combining both constraints, it would appear that the library has to be made available at a higher level than individual exercises, but not at the level of the judge, i.e., at the level of a course.

## Currently Considered Options
Our current idea would be to extend [the Scheme Dodona judge](https://docs.dodona.be/nl/references/judges/#scheme) with the logic to consider an extra `libraries` option to the `dirconfig.json` configuration of a course repository. Using the `libraries` option, course developers could then specify which libraries their exercises depend on. The libraries could be identified in one of the following ways, each with their own drawbacks:
- as a globally unique name **identifying a library that is shipped with the judge**. This is the easiest, but still entails that a _course-specific_ library becomes part of the shared Dodona judge.
- as **a path into a course-level folder** containing the libraries for that course. We're not sure what the best folder for this is. Neither `./media` nor `./public` really seem to be intended for source code. Are these folders even visible to the judge's `run` script? 
- as **a path into an exercise-level folder** containing a copy of the libraries for that exercise. Each exercise in the course would then have to contain an (identical) copy of the library. This is not a critical problem (the library we're targeting takes only 185KB spread across 73 files), but it feels like the wrong approach: it would be very inconvenient for the course maintainer to keep all copies of the library up to date as it is spread across all exercise folders.

We are leaning towards the middle option, as that one makes most sense, but that requires help from the Dodona developers: we'd need some place to put the library in [the repository directory structure](https://docs.dodona.be/en/references/repository-directory-structure/).

## Questions
Does anyone have a solution or insights in the best way to approach this? Does this sound like a more general problem that can be tackled by Dodona by, e.g., introducing a `./libraries` folder into the [repository directory structure](https://docs.dodona.be/en/references/repository-directory-structure/), or is this unique to our situation (and hence to be fixed by a "workaround" rather than a systemic solution)?

Thanks for your feedback.

**Comment by bmesuere** [comment]
We don't have a solution in place for "files used throughout one course". The most straightforward way would be to include the file in each exercise directory or include it once and use symlinks (although I'm not sure how git handles these).

We do have courses/exercises that require specific libraries. In these cases, we added the libraries to the docker image (so not the judge, but the image where it runs). An exercise can specify the image that needs to be used so this way, you could create a specific image for your course (e.g. by extending the existing image). We used this approach for a computational biology course that required additional python libraries. However, in practice, I think we now use the "big" image for all Python exercises.

**Reply by m0n0chr0m3** [reply]
Thanks Bart, that makes sense.

We considered using git submodules as a git-compatible way of having "symlinks", but having one submodule reference per exercise still seems excessive.

With "the docker images" do you mean [these](https://github.com/dodona-edu/docker-images)? I.e., do you suggest that I'd create a new `dodona-scheme-for-one-specific-course.dockerfile` in which the library comes pre-installed, instead of changing the existing `dodona-scheme.dockerfile` to support the new course ([as I started doing here](https://github.com/m0n0chr0m3/dodona-docker-images/commit/68ca49efe706f4599ecf247bd210683e077b830e))? That would indeed simplify things greatly for me, but that would pollute the shared `dodona-edu/docker-images` repo. It goes counter to what you seem to be aiming for with the move to TESTeD.

**Reply by bmesuere** [reply]
Yes, I meant those and exactly in the way you did. I'm ok with a separate image for your course, but if this extra library doesn't get in the way of other scheme exercises, the pragmatic solution would be to add it to the main scheme image (maybe with a comment of why it's there). So feel free to submit a PR to the images repo.
